### ISSUE TITLE
BERT has a non deterministic behaviour

### ISSUE THREAD

**Issue by RodSernaPerez** [google-research/bert#583]
I am using the BERT implementation in https://github.com/google-research/bert for feature extracting and I have noticed a weird behaviour which I was not expecting: if I execute the program twice on the same text, I get different results. I need to know if this is normal and why this happens in order to treat this fact in one or another way. Why is the reason for this? Aren't neural networks deterministic algorithms?

**Comment by Vibha111094** [comment]
I am also facing the same issue:(

**Comment by kenkrige** [comment]
I have just finished a kaggle competition where I made extensive use of BERT. It certainly is a little unstable in fine tuning and susceptible to random seed variations, especially with small data sets.

However, extracting features from a pre-trained model without changing the weights should yield the same output every time. There must be something wrong, but I am not entirely familiar with the code for feature extraction. Are you sure dropout is disabled in the code you are using to do feature extraction. If dropout was mistakenly left at the default training value of 0.9 (keep probability) then it would cause some variation in output.

**Comment by eikevons** [comment]
@RodSernaPerez How did you extract the features? Did you use `extract_features.py`? Can you share your code/command line?

**Comment by DiggiDon** [comment]
@RodSernaPerez, @Vibha111094 I had same issue too. My mistake was in `init_checkpoint`. The path ALWAYS should be ***/bert_model.ckpt, despite the fact, that you don't have the bert_model.ckpt file. For example, I have only bert_model.ckpt.data*** and bert_model.ckpt.index files.
Earlier I used ***.index file in`init_checkpoint` (just because there is no errors raised).

**Comment by dylstuart** [comment]
I have the same issue, I've removed dropout from the model, set my random seed with tf.random.set_random_seed() (Just in case), and I don't have the same mistake as @DiggiDon, so I'm at a loss. Anything else to try?

**Comment by sravanthi2019** [comment]
Is there a fix available for this because I am currently running a TFBertForSequenceClassification for classification and it gives non determinististic outcomes. I have written my code in tensorflow framework. Looks like the fix is placing model.eval() during evaluation operation as per https://github.com/google-research/bert/issues/583.
